python_modules/dagster/dagster/_core/definitions/assets_job.py

```diff
@@ -63,65 +63,62 @@ def is_base_asset_job_name(name: str) -> bool:
+        executable_assets_by_partitions_def[asset.partitions_def].append(asset)
+    # sort to ensure some stability in the ordering
+    all_partitions_defs = sorted(
+        [p for p in executable_assets_by_partitions_def.keys() if p], key=repr
+    )
+
+    if len(all_partitions_defs) == 0:
         return [
             build_assets_job(
                 name=ASSET_BASE_JOB_PREFIX,
-                assets=assets,
+                assets_to_execute=executable_assets,
+                other_assets=unexecutable_assets,
                 asset_checks=asset_checks,
-                source_assets=source_assets,
                 executor_def=executor_def,
                 resource_defs=resource_defs,
             )
         ]
     else:
-        unpartitioned_assets = assets_by_partitions_def.get(None, [])
-        partitioned_assets_by_partitions_def = {
-            k: v for k, v in assets_by_partitions_def.items() if k is not None
-        }
+        unpartitioned_executable_assets = executable_assets_by_partitions_def.get(None, [])
         jobs = []
 
-        # sort to ensure some stability in the ordering
-        for i, (partitions_def, assets_with_partitions) in enumerate(
-            sorted(partitioned_assets_by_partitions_def.items(), key=lambda item: repr(item[0]))
-        ):
+        # all partition base jobs contain all unpartitioned assets
+        for i, partitions_def in enumerate(all_partitions_defs):
+            partitioned_executable_assets = executable_assets_by_partitions_def[partitions_def]
+            assets_to_execute = [*partitioned_executable_assets, *unpartitioned_executable_assets]
             jobs.append(
                 build_assets_job(
                     f"{ASSET_BASE_JOB_PREFIX}_{i}",
-                    assets=[*assets_with_partitions, *unpartitioned_assets],
-                    source_assets=[*source_assets, *assets],
+                    assets_to_execute=assets_to_execute,
+                    other_assets=[
+                        *(asset for asset in executable_assets if asset not in assets_to_execute),
+                        *unexecutable_assets,
+                    ],
                     asset_checks=asset_checks,
                     resource_defs=resource_defs,
                     executor_def=executor_def,
-                    # Only explicitly set partitions_def for observable-only jobs since it can't be
-                    # auto-detected from the passed assets (which is an empty list).
-                    partitions_def=partitions_def if len(assets_with_partitions) == 0 else None,
                 )
             )
         return jobs
 
 
 def build_assets_job(
     name: str,
-    assets: Sequence[AssetsDefinition],
-    source_assets: Optional[Sequence[Union[SourceAsset, AssetsDefinition]]] = None,
+    assets_to_execute: Sequence[Union[AssetsDefinition, SourceAsset]],
+    other_assets: Optional[Sequence[Union[SourceAsset, AssetsDefinition]]] = None,
```

Comment:
We should rename these in another PR and let's come up with a better name than `other_assets`.

I also think we should introduce a typealias for `Union[AssetsDefinition, SourceAsset]` per our offline discussion for all the places where we will end up codemoding it back to `AssetsDefinition` only.